### PR TITLE
Mark all possible public methods leading to heap pollution with `@SafeVarargs` and make them final

### LINKED ISSUE #2161: Using softAssertions with extracting leads to warning related to unchecked generics array creation for varargs parameter
#### Summary

When I use `SoftAssertions` combined with `extracting`, my IDE complains with the following warning

> Unchecked generics array creation for varargs parameter

When I use the same code with standard `Assertions`, the warning is not there.

#### Example

With warning

```java
    TolkienCharacter legolas = TolkienCharacter.of("Legolas", 1000, ELF);

    softly.assertThat(legolas)
          .extracting(TolkienCharacter::getName, TolkienCharacter::getAge)
          .contains("Legolas", 1000);
```

Without warning

```java
    TolkienCharacter legolas = TolkienCharacter.of("Legolas", 1000, ELF);

    assertThat(legolas)
          .extracting(TolkienCharacter::getName, TolkienCharacter::getAge)
          .contains("Legolas", 1000);
```

### PR DESCRIPTION
**Pull request by filiphr**
Add intermediate protected methods for proxying that will be used by ByteBuddy to perform the actual delegation.

I took the liberty to remove some of the now not needed `@SuppressWarnings`.

With this, we can actually get rid of the different `ProxyXXXAssert`(s). I was not sure if you want me to do it now or not.
An additional improvement can also be the removal of the different `satisfiesAnyOf(Consumer<ACTUAL> assertions1, Consumer<ACTUAL> assertions2)` methods in the `AbstractAssert` and replace it with one with a varags.

@jgiovaresco, sorry that I did this without asking you first. I was eager to play a bit with ByteBuddy a bit. 

@joel-costigliola I am completely fine in giving the entire contribution about this to @jgiovaresco.

#### Check List:
* Fixes #2161
* Unit tests :  NA
* Javadoc with a code example (on API only) : NA

### CODE REVIEW

**Comment by scordio** [comment]
Thanks @filiphr, the idea is really nice.

> With this, we can actually get rid of the different `ProxyXXXAssert`(s). I was not sure if you want me to do it now or not.

I vote for it, those classes were already a limitation. It could be done in a separate PR, though.

> @joel-costigliola I am completely fine in giving the entire contribution about this to @jgiovaresco.

I think it would be fair to have the merge with a [co-authored commit](https://docs.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors).

**Comment by joel-costigliola** [comment]
That looks good, thanks @filiphr and @jgiovaresco.

File: src/main/java/org/assertj/core/api/AbstractClassAssert.java [hunk 1]
```diff
@@ -317,7 +317,12 @@ public SELF isPackagePrivate() {
    * @throws AssertionError if {@code actual} is {@code null}.
    * @throws AssertionError if the actual {@code Class} doesn't contains all of these annotations.
    */
-  public SELF hasAnnotations(@SuppressWarnings("unchecked") Class<? extends Annotation>... annotations) {
+  @SafeVarargs
+  public final SELF hasAnnotations(@SuppressWarnings("unchecked") Class<? extends Annotation>... annotations) {
+    return hasAnnotationsForProxy(annotations);
+  }
+
+  protected SELF hasAnnotationsForProxy(Class<? extends Annotation>[] annotations) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractClassAssert.java:325]: would be good to add a comment explaining why we have the `xxxForProxy` methods, if you feel duplicating the comment n times is a bit too much I'm happy to have it at the class level.

**Comment by jgiovaresco** [comment]
@filiphr Good job 👏 
I couldn't have done it as well 💪

File: src/main/java/org/assertj/core/api/AbstractClassAssert.java (hunk 1 above)
Comment (filiphr) [reply src/main/java/org/assertj/core/api/AbstractClassAssert.java:329]: Makes sense. I've added a comment to all methods. Please have a look at it and let me know if it is enough

**Comment by filiphr** [comment]
>I vote for it, those classes were already a limitation. It could be done in a separate PR, though.

Yes I agree, separate PR is better for it. I will work on one that will remove all the `ProxyXXXAssert`(s).

**Comment by joel-costigliola** [comment]
Thanks all! Really appreciate the good vibe :)

**Comment by joel-costigliola** [comment]
Integrated thanks @filiphr!
I also commited the removal of `ProxyXXXAssert`

**Comment by filiphr** [comment]
Thanks a lot @joel-costigliola

**Pull request closed without merging**